runtime_builders/publish_runtime.py

```diff
@@ -0,0 +1,161 @@
+
+LANGUAGES = [
+    'java',
+    'python',
+    'ruby',
+    'nodejs',
+    'golang',
+    'dotnet',
+    'php'
+]
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--infile', '-i', help='file(s) to read')
+    parser.add_argument('--bucket', '-b',
+                        help='GCS bucket to publish runtime to',
+                        default='runtime-builders')
+    parser.add_argument('--language', '-l',
+                        help='the language associated with this builder')
+    args = parser.parse_args()
+
+    if args.infile is None:
+        logging.error('Please specify templated input YAML file.')
+        return 1
+    if args.language is None:
+        logging.error('Please specify language to '
+                      'associate this builder with.')
+        return 1
+    if args.language not in LANGUAGES:
+        logging.error('Invalid language \'{0}\' specified! '
+                      'Options:'.format(args.language))
+        logging.error(LANGUAGES)
+        return 1
+
+    templated_file = _resolve_tags(args.infile)
+    _publish_to_gcs(templated_file, args.language, args.bucket)
+
+
+def _resolve_tags(config_file):
+    """
+    Given a templated YAML cloudbuild config file, parse it, resolve image tags
+    on each build step's image to the corresponding digest, and write new
+    config with fully qualified images to temporary file for upload to GCS.
+
+    Keyword arguments:
+    config_file -- string representing path to
+    templated cloudbuild YAML config file
+
+    Return value:
+    path to temporary file containing fully qualified config file, to be
+    published to GCS.
+    """
+    with open(config_file, 'r') as infile:
+        try:
+            config = yaml.round_trip_load(infile)
+
+            for step in config.get('steps'):
+                image = step.get('name')
+                templated_step = _resolve_tag(image)
+                step['name'] = templated_step
+
+            s = yaml.round_trip_dump(config)
+
+            fd, ofile = tempfile.mkstemp()
+            with open(ofile, 'w') as outfile:
+                outfile.write(s)
+            outfile.close()
+            os.close(fd)
+            return ofile
+
+        except yaml.YAMLError as e:
+            logging.error(e)
+        finally:
+            infile.close()
+
+
+def _resolve_tag(image):
+    """
+    Given a path to a tagged Docker image in GCR, replace the tag with its
+    corresponding sha256 digest.
+    """
+    if image is None:
+        raise Exception('Please provide image')
+
+    if ':' not in image:
+        logging.info('Image does not contain tag: defaulting to "latest"')
+        base_image = image
+        target_tag = 'latest'
+    else:
+        base_image = image.split(':')[0]
+        target_tag = image.split(':')[1].strip('${}')
+
+    command = ['gcloud', 'beta', 'container', 'images',
+               'list-tags', base_image, '--format=json']
```

Comment:
update: `describe` requires that the active account has read permissions to the project associated with the image. because we can't guarantee this at runtime, going to go back to using `list-tags`.